### ISSUE TITLE
\yii\mailer\BaseMailer can't be extended due to bad design choices

### ISSUE THREAD

**Issue by creocoder** [yiisoft/yii2#9260]
The problem is class can not be extended due to bad design. Example:

``` php
class MyMailer extends \yii\mailer\BaseMailer
{
    // ...
    public function compose($view = null, array $params = [])
    {
        // my own logic, but i want use $this->render() in my implementation too
        $this->render($view['html'], $params, $this->htmlLayout);
    }
}
```

In child classes `$message` in layouts will always be `null` and you can't do anything with that. Look at original class compose() method:

``` php
    public function compose($view = null, array $params = [])
    {
        if (!array_key_exists('message', $params)) {
            $params['message'] = $message;
        }

        // ...
        $this->_message = $message;
        // render call
        $this->_message = null;
        // ...
    }

    public function render($view, $params = [], $layout = false)
    {
        // $this->_message used there and you can't set its value from child classes
    }
```

Its what i call voodoo magic which make this class methods complettely unextensible.

**Comment by klimov-paul** [comment]
Relates #4147

**Comment by klimov-paul** [comment]
Since solution can be achieved only via method signature change or view internal code signature change, the issue is unlikely to be resolved until version 2.1.

**Comment by creocoder** [comment]
@klimov-paul 

> Since solution can be achieved only via method signature change

Why? We can do it for example `render($view, $params = [], $layout = false, $message = null)`.

It if we want to solve trouble in 2.0. But in general main question is: why anyone need message inside layout?! Message object inside concrette view really make sense, message inside layout does not make sense at all. And even if user needs it in layout he can transfer it via view params.

**Comment by klimov-paul** [comment]
> We can do it for example render($view, $params = [], $layout = false, $message = null)

This way it will trigger E_NOTICE in case someone overrides this method as you do.

> But in general main question is: why anyone need message inside layout?

See #3992

**Comment by creocoder** [comment]
@klimov-paul Yes, you right. What you think about `$message` necessity inside layout?

**Comment by creocoder** [comment]
@klimov-paul 

> See #3992

But it closed. And than it still implemented without any reason. Do you see any reasons to have `$message` instance inside layout? Can you show any example how why this may be useful?

**Comment by klimov-paul** [comment]
Resolved by #13809
